lib/tlog/play.c

```diff
@@ -64,6 +64,234 @@ tlog_play_io_sighandler(int signum)
+            goto cleanup;
+        }
+        for (i = 0; (int)i < (int)json_object_array_length(obj); i++) {
+             str_list[i] = json_object_get_string(
+                            json_object_array_get_idx(obj, i));
+             if (!tlog_journal_match_sym_is_valid(str_list[i])) {
+                 grc = TLOG_RC_FAILURE;
+                 tlog_errs_pushf(
+                     perrs,
+                     "Systemd journal match symbol #%zu \"%s\" "
+                     "is invalid",
+                     i + 1, str_list[i]);
+                 goto cleanup;
+             }
+         }
+     }
+
+    /* Create the reader */
+    grc = tlog_journal_json_reader_create(&reader,
+                                          (uint64_t)since * 1000000,
+                                          (uint64_t)until * 1000000,
+                                          str_list);
+    if (grc != TLOG_RC_OK) {
+        tlog_errs_pushc(perrs, grc);
+        tlog_errs_pushs(perrs,
+                        "Failed creating the systemd journal reader");
+        goto cleanup;
+    }
+
+    *preader = reader;
+    reader = NULL;
+    grc = TLOG_RC_OK;
+
+cleanup:
+    free(str_list);
+    tlog_json_reader_destroy(reader);
+    return grc;
+}
+#endif
+
+/**
+ * Create a file JSON message reader according to configuration.
+ *
+ * @param perrs         Location for the error stack. Can be NULL.
+ * @param preader       Location for the created reader pointer. Not modified
+ *                      in case of error.
+ * @param conf          Configuration JSON object.
+ *
+ * @return Global return code.
+ */
+ static tlog_grc
+ tlog_play_create_file_json_reader(struct tlog_errs **perrs,
+                                   struct tlog_json_reader **preader,
+                                   struct json_object *conf)
+{
+     tlog_grc grc;
+     const char *str;
+     int fd = -1;
+     struct tlog_json_reader *reader = NULL;
+     struct json_object *obj;
+
+     assert(preader != NULL);
+     assert(conf != NULL);
+
+     /* Get the file path */
+     if (!json_object_object_get_ex(conf, "path", &obj)) {
+         tlog_errs_pushs(perrs, "Log file path is not specified");
+         grc = TLOG_RC_FAILURE;
+         goto cleanup;
+     }
+     str = json_object_get_string(obj);
+
+     /* Open the file */
+     fd = open(str, O_RDONLY);
+     if (fd < 0) {
+         grc = TLOG_GRC_ERRNO;
+         tlog_errs_pushc(perrs, grc);
+         tlog_errs_pushf(perrs, "Failed opening log file \"%s\"", str);
+         goto cleanup;
+     }
+
+     /* Create the reader, letting it take over the FD */
+     grc = tlog_fd_json_reader_create(&reader, fd, true, 65536);
+     if (grc != TLOG_RC_OK) {
+         tlog_errs_pushc(perrs, grc);
+         tlog_errs_pushs(perrs, "Failed creating file reader");
+         goto cleanup;
+     }
+     fd = -1;
+
+     *preader = reader;
+     reader = NULL;
+     grc = TLOG_RC_OK;
+
+cleanup:
```

Comment:
The body of this function above this line has one extra space of indent. Please fix.